### PR TITLE
fix(builder): step references wrapped in quotes no longer turn into raw text

### LINKED ISSUE #14815: [BUG] Step reference wrapped in quotes loses its chip and renders as raw text on builder remount
## Problem

ShortText value like `"{{step_4['output']['result']}}"` shows the reference chip while editing, but once the step settings remount (step switch on cloud/main; panel close+reopen on 0.88.0; page reload) the chip renders as raw `{{...}}` text. Display only: saved value intact, runtime resolves correctly.

Quoted references are legitimate input, e.g. Google Drive Search `fullText contains '"exact phrase"'`.

Worse than reported: one unpaired `"` anywhere earlier in the value kills chips for every later reference in that field.

Regression since 0.85.0 (PR 12444, formula functions). Live-verified on 0.88.0.

Ref: Pylon 5621

## Steps to reproduce

1. Any ShortText input (e.g. Google Drive Search → Value): insert a step reference chip, surround it with `"` quotes
2. Click another step, click back (on 0.88.0: close the step panel and reopen)
3. Chip replaced by raw `"{{step_4['output']['result']}}"` text

## Root cause

`packages/web/src/app/builder/piece-properties/text-input-with-mentions/text-input-utils.ts:126-141` — `tokenizeExpression`'s text-accumulation loop enters string mode on any `"`/`'` (added to protect `)`/`;` inside quoted function args) and consumes `{{` without breaking out → no variable token → mention chip never re-created on re-parse. Remount trigger: `StepSettingsProvider` is keyed by step name (`packages/web/src/app/builder/index.tsx:230`), so any step switch re-parses the saved text.

## Fix

In string mode still break on `{{` (or track strings only when `fnDepth > 0`). Either keeps the function-arg protection.

## Verification

* 5 passing unit tests: `packages/web/test/app/builder/piece-properties/text-input-with-mentions/quoted-reference-chip-claim-verify.test.ts` (quoted loses chip, unquoted control, unpaired-quote spread, single quotes, round-trip lossless)
* Live repro on 0.88.0 docker image — screenshots in comments

### PR DESCRIPTION
**Pull request by majewskibartosz**
Fixes [GIT-1752](https://linear.app/activepieces/issue/GIT-1752)
Closes #14815

## Problem

1. A step reference wrapped in quotes (`"{{step_4['output']['result']}}"`) shows its data chip while editing, but after the step panel remounts (step switch, panel reopen, reload) it renders as raw `{{...}}` text. Display-only; runtime resolution unaffected. Quoted references are valid input, e.g. Google Drive Search `fullText contains '"exact phrase"'`.
2. One unpaired `"` earlier in a value killed the chips of every later reference in that field.
3. Regression since 0.85.0 (#12444): the tokenizer's string-literal mode consumed `{{` without emitting a variable token.

## Fix

- `text-input-utils.ts` — `tokenizeExpression` now recognizes `{{...}}` inside the text-accumulation loop and emits the variable token inline, so string state survives across a reference (a quoted function argument containing a reference keeps `)`/`;` protection and round-trips losslessly). The duplicate top-of-loop `{{` block is deleted; one mechanism handles closed, unclosed, and in-string references.
- `stringClosesWithin` guard: a `{{` opened inside a quoted string whose `}}` lies past the closing quote stays literal text, preventing the mention from swallowing the quote and corrupting the value on re-serialization.

## Verification

- Regression tests red-first: 4 mention assertions fail on base commit, pass with fix (`text-input-utils.test.ts`)
- Round-trip suite: 7 inputs serialize back byte-identical, including refs inside quoted function args and the `concat("{{a"}}; lower(x))` boundary case
- Full `packages/web` suite: 476 passed; `chunk-reducer.test.ts` fails identically on the pristine base commit (`window is not defined`, pre-existing, unrelated)
- `npm run lint-dev`: 32/32 clean

Repro: reproduced live on a 0.88.0 docker stack, then drove the fixed build against the same backend and data — chip survives mount, panel reopen, step switch, and reload — full steps in the Reproduction block

Visual: driven states on a real running instance — quoted chip on fresh mount, after panel close+reopen, after step switch, after full reload, quote removal and re-add (saved value verified via API each time); light theme (change is not theme-sensitive)

Other affected paths: checked — the sibling sweep found no second `{{`-tokenizer (formula evaluator never sees raw references). Three pre-existing round-trip corruption cases in the same loop fail identically on base and are left unfixed here: lone apostrophe in an unquoted formula arg, newline inside function args, escaped backslash before a closing quote
- Two accepted behavior changes: literal `{{ ... }}` inside quotes now renders a `(Missing)` chip (matches engine interpolation semantics, round-trip stays lossless), and quoted refs inside formula args no longer trigger static type hints (runtime type is genuinely unknowable)
- Size: product diff grew past the initial ~2-line estimate because review probes proved the naive break corrupts quoted-function-arg values; the inline-emission restructure plus the string-boundary guard is the smallest shape that fixes the chip without any lossy round-trip

<details>
<summary>Plan & reasoning</summary>

## Plan

- Root cause pinned in [GIT-1752](https://linear.app/activepieces/issue/GIT-1752) internal comment: string mode in `tokenizeExpression` swallowed `{{`
- Chosen shape: emit variable tokens from inside the accumulation loop so `inString` survives across a reference; delete the outer duplicate `{{` block
- Footprint estimate: 2 files ~2 lines; actual: 2 files, +71/−15 (growth explained by Size line)
- Remount trigger (`StepSettingsProvider` keyed by step name, `index.tsx:230`) intentionally untouched: remount-on-switch is intended; fixing the parser fixes every re-parse path (paste, undo, import), not just this trigger

## Non-happy scenarios considered

- Reference inside quoted function arg followed by `)`/`;` → inline emission keeps string protection; round-trip test locks it
- `{{` inside a string whose `}}` lies past the closing quote → `stringClosesWithin` keeps it literal; round-trip test locks it
- Unclosed `{{` while typing → consumed as literal text, no tokenizer spin (existing tests still pass)
- Escaped quotes inside references → outer scan is quote-agnostic, consistent with engine interpolation

## Alternatives rejected

- Break out of string mode on `{{` (one-liner) → resets `inString`, corrupts quoted-function-arg values on next save (proven by probe)
- Track strings only when `fnDepth > 0` → leaves quoted refs inside function args chipless, and fn-name detection after a ref still corrupts top-level values
- Change panel remount keying → hides one trigger path, leaves paste/undo/import broken
</details>

<details>
<summary>Reproduction</summary>

## Environment

- `activepieces/activepieces:0.88.0` + `postgres:14` + `redis:7` via docker compose, app on `localhost:8099`, `AP_EXECUTION_MODE=UNSANDBOXED`, telemetry off
- Fixed build verified by running the branch's `packages/web` vite dev server proxied to the same backend

## Harness

`docker-compose.yml` (compose stack) and `seed.sh` (signs up a user, creates platform/project, seeds a flow: Catch Webhook trigger → Google Drive Search with `query` = `"{{trigger['body']['searchTerm']}}"`):

```yaml
services:
  app:
    image: activepieces/activepieces:0.88.0
    ports: ['8099:80']
    depends_on: [postgres, redis]
    environment:
      AP_ENGINE_EXECUTABLE_PATH: dist/packages/engine/main.js
      AP_ENCRYPTION_KEY: e81f8754faa04a00beaf5b358ddacbcf
      AP_JWT_SECRET: repro-5634-jwt-secret
      AP_ENVIRONMENT: prod
      AP_FRONTEND_URL: http://localhost:80
      AP_POSTGRES_DATABASE: activepieces
      AP_POSTGRES_HOST: postgres
      AP_POSTGRES_PORT: '5432'
      AP_POSTGRES_USERNAME: postgres
      AP_POSTGRES_PASSWORD: ap
      AP_EXECUTION_MODE: UNSANDBOXED
      AP_REDIS_HOST: redis
      AP_REDIS_PORT: '6379'
      AP_TELEMETRY_ENABLED: 'false'
  postgres:
    image: postgres:14
    environment:
      POSTGRES_DB: activepieces
      POSTGRES_PASSWORD: ap
      POSTGRES_USER: postgres
  redis:
    image: redis:7
```

```bash
docker compose -p repro1752 up -d
./seed.sh    # prints the flow URL + login
```

seed.sh (throwaway credentials, local containers only):

```bash
#!/usr/bin/env bash
set -euo pipefail
BASE=http://localhost:8099/api/v1
SIGNUP=$(curl -s -X POST $BASE/authentication/sign-up -H 'Content-Type: application/json' -d '{"email":"repro@example.com","password":"Repro5634!x","firstName":"Repro","lastName":"Bot","trackEvents":false,"newsLetter":false}')
ONBOARDING_TOKEN=$(echo "$SIGNUP" | python3 -c 'import json,sys;print(json.load(sys.stdin)["token"])')
PLATFORM=$(curl -s -X POST $BASE/platforms -H "Authorization: Bearer $ONBOARDING_TOKEN" -H 'Content-Type: application/json' -d '{"name":"Repro Platform"}')
TOKEN=$(echo "$PLATFORM" | python3 -c 'import json,sys;print(json.load(sys.stdin)["token"])')
PROJECT=$(echo "$PLATFORM" | python3 -c 'import json,sys;print(json.load(sys.stdin)["projectId"])')
FLOW=$(curl -s -X POST $BASE/flows -H "Authorization: Bearer $TOKEN" -H 'Content-Type: application/json' -d "{\"projectId\":\"$PROJECT\",\"displayName\":\"Quoted chip repro\"}" | python3 -c 'import json,sys;print(json.load(sys.stdin)["id"])')
curl -s -X POST $BASE/flows/$FLOW -H "Authorization: Bearer $TOKEN" -H 'Content-Type: application/json' -d '{"type":"UPDATE_TRIGGER","request":{"name":"trigger","type":"PIECE_TRIGGER","displayName":"Catch Webhook","valid":true,"settings":{"pieceName":"@activepieces/piece-webhook","pieceVersion":"0.1.40","triggerName":"catch_webhook","input":{"authType":"none"},"propertySettings":{}}}}' > /dev/null
curl -s -X POST $BASE/flows/$FLOW -H "Authorization: Bearer $TOKEN" -H 'Content-Type: application/json' -d '{"type":"ADD_ACTION","request":{"parentStep":"trigger","action":{"name":"step_1","type":"PIECE","displayName":"Search","valid":false,"settings":{"pieceName":"@activepieces/piece-google-drive","pieceVersion":"0.8.3","actionName":"search-folder","input":{"queryTerm":"fullText","operator":"contains","query":"\"{{trigger['"'"'body'"'"']['"'"'searchTerm'"'"']}}\""},"propertySettings":{},"errorHandlingOptions":{}}}}' > /dev/null
echo "flow: http://localhost:8099/projects/$PROJECT/flows/$FLOW  (login: repro@example.com / Repro5634!x)"
```

## Trigger

Open the seeded flow, open the Search step (0.88.0: value renders as raw text immediately, the bug). On a build with this fix, the same open renders `"`+chip+`"`. The customer's gesture (step switch) reproduces on current main because `StepSettingsProvider` is keyed by step name; on 0.88.0 the panel must be closed and reopened.

## Results

- 0.88.0 (unfixed): quoted value renders as raw `"{{trigger['body']['searchTerm']}}"` text on every panel mount; API read confirms the saved value itself never changes (display-only)
- Branch build against same backend/data: chip renders on mount, panel reopen, step switch, reload; removing the quotes in the UI saves the bare reference; round-trip byte-identical each time
</details>

### Breaking change?  (required — CI fails if this is left unedited)

- [x] no — reviewed, not breaking
- [ ] yes — technical (removed/renamed API field or endpoint, dropped column, new required field, removed/required env var)
- [ ] yes — functional (default/limit/behaviour change, new self-hosted setup step)

### Security impact?  (required — CI fails if this is left unedited)

- [x] no — reviewed, no security impact
- [ ] yes — security-sensitive (call out the risk and mitigation in the description above)

### CODE REVIEW

**Comment by majewskibartosz** [comment]
## Visual verification

Real running instance (0.88.0 docker backend + this branch's frontend), driven via browser automation. Saved value throughout: `"{{trigger['body']['searchTerm']}}"`, verified byte-identical via API after every step.

| Before (base behavior) | After (this fix) |
|---|---|
| before: raw text after reopen | after: chip persists |

While editing (both before and after — the chip is only lost on re-parse pre-fix):

editing state

Additional driven states on the fixed build: after panel close+reopen (reopen), after full page reload (reload), step switch away/back, quote removal (saved value drops quotes, reference intact) and quote re-add. Light theme; the change is not theme-sensitive.

File: packages/web/src/app/builder/piece-properties/text-input-with-mentions/text-input-utils.ts [hunk 1]
```diff
@@ -126,6 +129,29 @@ function tokenizeExpression(expr: string, allowBroken: boolean): ExprToken[] {
...
+          continue;
+        }
+        if (
+          inString &&
+          stringClosesWithin({ expr, from: i + 2, to: end, quote: inString })
+        ) {
+          text += ch;
+          i++;
+          continue;
+        }
```
[the remark this answers is not in the document]
Comment (majewskibartosz) [reply packages/web/src/app/builder/piece-properties/text-input-with-mentions/text-input-utils.ts]: Valid finding. Fixed in 27e683ea57: replaced the bail-to-text guard (`stringClosesWithin`) with `stringStateAfter`, which emits the reference and carries string state through the span — quotes inside the reference toggle the tracker instead of suppressing the chip. `"{{step_4["output"]["result"]}}"` now renders its chip (even quote count keeps the outer string open, the trailing `"` closes it), and the corruption case the guard existed for (`concat("{{a"}}; lower(x))`, odd quote count closes the string so `;`/`)` stay structural) still round-trips byte-identical. Both are locked in `text-input-utils.test.ts` (mention + round-trip suites), plus the out-of-string odd-quote ref `upper({{step_1["a'b"]}}; x)` to pin that references outside strings stay opaque.

File: packages/web/src/app/builder/piece-properties/text-input-with-mentions/text-input-utils.ts [hunk 2]
```diff
@@ -67,27 +67,36 @@ type ExprToken =
...
+  to: number;
+  state: '"' | "'";
+}): '"' | "'" | null {
+  let current: '"' | "'" | null = state;
+  for (let j = from; j < to; j++) {
+    const c = expr[j];
+    if (current) {
+      if (c === current && expr[j - 1] !== '\\') current = null;
+    } else if (c === '"' || c === "'") {
+      current = c;
```
[the remark this answers is not in the document]
Comment (majewskibartosz) [reply packages/web/src/app/builder/piece-properties/text-input-with-mentions/text-input-utils.ts]: Valid. Fixed in c05912b648: the opening transition in stringStateAfter now skips escaped quotes, same as the closing transition, so `\'` inside a bracket key never opens a phantom string. Your exact input round-trips byte-identical and is locked in the round-trip suite (upper('pre {{step_1['output']['a\'b']}} and lower(x)') keeps its formula suffix; the tail stays literal quoted content).

**Review by Louai-Zokerburg — APPROVED** [review]
